bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -530,52 +598,79 @@ public synchronized void start() throws MqttException, ConfigurationException {
         if (connectionState() != MqttConnectionState.DISCONNECTED) {
             return;
         }
-
         // Ensure the reconnect strategy is started
         if (reconnectStrategy != null) {
             reconnectStrategy.start();
         }
 
-        // Storage
-        String tmpDir = System.getProperty("java.io.tmpdir") + "/" + host;
-        MqttClientPersistence dataStore = new MqttDefaultFilePersistence(tmpDir);
+        // Close client if there is still one existing
+        if (client != null) {
+            try {
+                client.close();
+            } catch (org.eclipse.paho.client.mqttv3.MqttException ignore) {
+            }
+            client = null;
+        }
+
+        // Perform the connection attempt
+        isConnecting = true;
+        connectionObservers.forEach(o -> o.connectionStateChanged(MqttConnectionState.CONNECTING, null));
+
+        client = createAndConnectClient();
 
+        ScheduledExecutorService executor = timeoutExecutor;
+        if (executor != null && connectionToken != null) {
+            final ScheduledFuture<?> timeoutFuture = this.timeoutFuture.getAndSet(
+                    executor.schedule(() -> connectionCallbacks.onFailure(connectionToken, new TimeoutException()),
+                            timeout, TimeUnit.MILLISECONDS));
+            if (timeoutFuture != null) {
+                timeoutFuture.cancel(false);
+            }
+        }
+    }
+
+    /**
+     * Encapsulates the creation of the paho MqttAsyncClient and client connection.
+     *
+     * @param dataStore The datastore to save qos!=0 messages until they are delivered.
+     * @return Returns a valid and connecting MqttAsyncClient
+     * @throws MqttException If an exception of the underlying library happens, this exception is thrown
+     * @throws ConfigurationException The configuration is not valid if this exception is thrown
+     */
+    protected MqttAsyncClient createAndConnectClient() throws MqttException, ConfigurationException {
         StringBuilder serverURI = new StringBuilder();
         serverURI.append((secure ? "ssl://" : "tcp://"));
         serverURI.append(host);
         serverURI.append(":");
         serverURI.append(port);
 
-        // Close client if there is still one existing
-        if (client != null) {
-            try {
-                client.close();
-            } catch (org.eclipse.paho.client.mqttv3.MqttException ignore) {
+        // Storage
+        Path tmpDir = Paths.get(ConfigConstants.getUserDataFolder());
```

Comment:
when running from the IDE my tmpDir is always `userdata` and not absolute.
This can be fixed by adding `toAbsolutePath()` to the `Paths#get` call. wdyt?

Reply:
The user dir is always absolute in production environment. I added this precheck to not make the class create directories in various test working dirs.

It is actually a bug in one of the tests if this method is not overwritten with a no-op.

Reply:
> The user dir is always absolute in production environment.

Why?

The code looks like:

```Java
public static final String DEFAULT_USERDATA_FOLDER = "userdata";

// ...

public static String getUserDataFolder() {
    String progArg = System.getProperty(USERDATA_DIR_PROG_ARGUMENT);
    if (progArg != null) {
        return progArg;
    } else {
        return DEFAULT_USERDATA_FOLDER;
    }
}
```

So for example if `progArg == null` then `userdata`  is returned.
This is not absolute.

Also the `USERDATA_DIR_PROG_ARGUMENT` property does not need to be an absolute path.

Reply:
> It is actually a bug in one of the tests if this method is not overwritten with a no-op.
tests are fine.

From my pov you can just create the folders on the user directory. @maggu2810 any objections?

Reply:
Hm, but tests shouldn't alter the working directory, should they? I only checked with my OH environment and the path was absolute, I thought that's the differentiation of testing Vs production systems.

Reply:
> From my pov you can just create the folders on the user directory. @maggu2810 any objections?

The user data folder needs to be a writable location.
Why do you think the user directory needs to be a writable one?
This has till now never be a requirement for an ESH based solution (AFAIK).
And why should we fill the user directory with ESH stuff at all? Isn't that the intention of the user data folder, that ESH runtime data is stored at that place?
This will also break solutions that runs multiple ESH instances as the same OS user.

> but tests shouldn't alter the working directory, should they

I agree that the CWD should not be changed at runtime.
Using absolute or relative should not care, because if the CWD is changed also the conversion to an absolute path will result into a wrong directory.
I just want to point out, that `getUserDataFolder` can return an relative or an absolute directory, that depends on the system properties.

So, I am fine using the path returned by `getUserDataFolder()` and create a subfolder `mqtt` in that place (perhaps we should use `io.transport.mqtt` to not use the same directory for different MQTT bundles some time).

Reply:
The test run well in the ESH IDE. No folders get created afaik.
But running ESH with io.transport.mqtt from the IDE results in the Exception for every new service instance I configure. So I just suggest to make the path absolute. Or do I overlook a problem here?

Reply:
so, to clear this up:

> So, we have to use getUserDataFolder() and not the user's home directory.

Agreed. My suggestion was misleading due to imprecise wording. I was also talking about `userdata` which is the default user data folder.

> If the test only succeeds only if this directory is absolute

as @davidgraeff mentioned, unit tests do override the corresponding `createAndConnectClient` method and do not create any folders. Its only the ESH execution in the IDE (which falls back to the default `userdata` relative path).
With the use of `toAbsolutePath()` we should be all golden.